### PR TITLE
[skip changelog] Updated the deprecated ioutil dependency

### PR DESCRIPTION
**Pull request by zachary-walters**
## Please check if the PR fulfills these requirements

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

- [x] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [x] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] `UPGRADING.md` has been updated with a migration guide (for breaking changes)

## What kind of change does this PR introduce?

This change updates the codebase to use alternatives to the ioutil dependency. [ioutil has been deprecated since go v1.16](https://go.dev/doc/go1.16#ioutil).

## What is the current behavior?

The current behavior is unchanged.

## What is the new behavior?

There is no new behavior

## Does this PR introduce a breaking change, and is [titled accordingly](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#breaking)?

No

## Other information

Please see go's official statement about ioutil's deprecation in v1.16: https://go.dev/doc/go1.16#ioutil

### CODE REVIEW

File: arduino/resources/helpers_test.go [hunk 1]
```diff
@@ -16,7 +16,7 @@
 package resources
 
 import (
-	"io/ioutil"
+	"os"
 	"net/http"
 	"net/http/httptest"
```
Comment (per1234) [inline arduino/resources/helpers_test.go:21]:
```suggestion
	"net/http"
	"net/http/httptest"
	"os"
```

Resolve the `go fmt` violation that causes the "Check Go" CI workflow run failure:

https://github.com/arduino/arduino-cli/actions/runs/4039005652/jobs/6943394036

**Review by per1234 — CHANGES_REQUESTED** [review]

File: legacy/builder/utils/utils.go [hunk 2]
```diff
@@ -443,12 +443,21 @@ func CopyDir(src string, dst string, extensions []string) (err error) {
...
+	infos := make([]fs.FileInfo, 0, len(entries))
 	for _, entry := range entries {
+		info, scopeErr := entry.Info()
+		if scopeErr != nil {
+			return
+		}
+		infos = append(infos, info)
+	}
+
+	for _, entry := range infos {
```
Comment (cmaglie) [inline legacy/builder/utils/utils.go:460]: Looks like we don't need to allocate the array here

```suggestion
	for _, dirEntry := range entries {
		entry, scopeErr := dirEntry.Info()
		if scopeErr != nil {
			return
		}
```

**Review by cmaglie — CHANGES_REQUESTED** [review]
HI @zachary-walter!

I've added a couple of nitpicks, but for the rest LGTM, we can merge it as soon as you fix them.
Thanks!

File: legacy/builder/ctags/ctags_parser_test.go [hunk 3]
```diff
@@ -11,12 +11,12 @@
 // a commercial license. Buying such a license is mandatory if you want to
 // modify or otherwise use the software for commercial activities involving the
 // Arduino software without disclosing the source code of your own applications.
-// To purchase a commercial license, send an email to license@arduino.cc.
+// To purchase a commercial license, send an email to license@arduino.c`c.
```
[the remark this answers is not in the document]
Comment (zachary-walters) [reply legacy/builder/ctags/ctags_parser_test.go]: @cmaglie 

Apologies here. This is a typo likely due to missing the esc key.

File: legacy/builder/utils/utils.go (hunk 2 above)
Comment (zachary-walters) [reply legacy/builder/utils/utils.go]: I have removed the array allocation block

**Review by cmaglie — APPROVED** [review]

**Pull request merged**